crates/types/src/dedup.rs

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) 2024 - Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use crate::identifiers::{LeaderEpoch, PartitionId};
+use crate::message::MessageIndex;
+use bytestring::ByteString;
+use std::cmp::Ordering;
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct DedupInformation {
+    pub producer_id: ProducerId,
+    pub sequence_number: DedupSequenceNumber,
+}
+
+impl DedupInformation {
+    pub fn cross_partition(producer_id: PartitionId, sequence_number: MessageIndex) -> Self {
+        DedupInformation {
+            producer_id: ProducerId::Partition(producer_id),
+            sequence_number: DedupSequenceNumber::Sn(sequence_number),
+        }
+    }
+
+    pub fn self_proposal(esn: EpochSequenceNumber) -> Self {
+        DedupInformation {
+            producer_id: ProducerId::self_producer(),
+            sequence_number: DedupSequenceNumber::Esn(esn),
+        }
+    }
+
+    pub fn ingress(producer_id: impl Into<ByteString>, sequence_number: MessageIndex) -> Self {
+        DedupInformation {
+            producer_id: ProducerId::Other(producer_id.into()),
+            sequence_number: DedupSequenceNumber::Sn(sequence_number),
+        }
+    }
+}
+
+static SELF_PRODUCER: ByteString = ByteString::from_static("INTERNAL_SELF_PRODUCER");
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub enum ProducerId {
+    Partition(PartitionId),
+    Other(ByteString),
```

Comment:
Because right now, we are also using `Other` when doing self-proposals (to store another dedup information apart from the shuffle dedup information which is identified via the `PartitionId`, note that sending messages to oneself currently also goes through the shuffle). We can distinguish these two cases if this becomes necessary.